New graphics for Control Options menu, Display Mode option and Health Indicator option

mercury501 opened the pull request:
Hi!

This PR contains 3 features:
Replacing the Buttons text in Control Options with custom icons, and new graphics in that menu
Display mode in Advanced Options
Health indicator in Game Options

Let me know if you need any changes!

Cheers!

Polymega commented:
You're a badass, mercury!! This was a massive undertaking but you pulled it through, like always. ❤️ 

Note that these features only work for 1.0 as of now. @AeroWidescreen may I ask if you'd consider helping us find equivalent addresses for 1.1 and DC?

mercury501 commented:
For future reference, addresses to dial in are:
[Here, for Health indicator](https://github.com/mercury501/Silent-Hill-2-Enhancements/blob/8e9a990c84d0585c48056aac8851a6511c3cdab7/Patches/OptionsMenuTweaks.cpp#L1047)
[Here, for display mode](https://github.com/mercury501/Silent-Hill-2-Enhancements/blob/8e9a990c84d0585c48056aac8851a6511c3cdab7/Patches/OptionsMenuTweaks.cpp#L1336)
[Here](https://github.com/mercury501/Silent-Hill-2-Enhancements/blob/8e9a990c84d0585c48056aac8851a6511c3cdab7/Patches/OptionsMenuTweaks.cpp#L749) for control options

AeroWidescreen commented:
I'll get started on this now.

mercury501 commented:
And here's the mes file to test with:
[option_msg_e.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/14705193/option_msg_e.zip)

mercury501 commented:
Thanks @AeroWidescreen for the addresses!

Polymega commented:
Love the teamwork! 😍 Thank you both so much.

Mercury did you do a quick test with 1.1 and DC after getting their addresses?

mercury501 commented:
Hi @Polymega, I still need to find the addresses for the Display Mode option, so I couldn't test on other versions yet

AeroWidescreen commented:
@Polymega We agreed to split the work, so my half is done, but Mercury already knows he can ask me for additional help if he needs it. Some of them can be tricky to find.

Polymega commented:
I'm not sure how to consistently repeat it, but there has been two times now where the game has launched using low resolution textures. When this happens, it can't be changed whenever the new Display Mode feature is in effect.

Polymega commented:
I'm working on getting translations for these new strings for the six, base languages. Once I get those, I'll share here to help with testing.

Polymega commented:
> I'm working on getting translations for these new strings for the six, base languages. Once I get those, I'll share here to help with testing.

I have translations for all the necessary strings now relating to the features Mercury is PR'ing here. Attached is a ZIP for the `r_menu_*.res` files: [r_menu.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/14910924/r_menu.zip)

elishacloud commented:
I fixed the issue with low res textures getting re-enabled.  I also added support for dynamically changing display modes on the fly and dynamically changing the health indicator.

@mercury501, I still see two "TODO" notes.  [Here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/pull/935/files#diff-9c12d0c188c16765dde0cd9c23c19b1957a148784fb5a24adffce9d5f4b67235R1070) and [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/pull/935/files#diff-9c12d0c188c16765dde0cd9c23c19b1957a148784fb5a24adffce9d5f4b67235R1579).  See below.  What is the status of these two items?

```c++
void PatchHealthIndicatorOption() // TODO find faulty address in 1.1 and DC
```

```c++
WriteJMPtoMemory(DisplayModeValueHighlightAddr, DisplayModeValueHighlight, 0x08); //TODO highlight wrong color, tied to inputs
```

Polymega commented:
I'm not sure if the latest Action Builds incorporate your commits for this PR, but I tried the latest build and the game still enables low res textures on resolution change:

### "High res" textures
![Screenshot 2024-04-20 13-10-01](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/27395260/1c658da1-c352-4f83-9327-e37eeb8d30bd)

### "Low res" textures
![Screenshot 2024-04-20 13-10-16](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/27395260/a460acc7-c56f-4167-8abb-c2079e372beb)

Low/high res textures won't take effect until you load a new room or reload the same room. To test, change the resolution, go back to the game, press F5 to Quick Save then F9 to Quick Load back into the room. This will reload the room.

Polymega commented:
*Note the high/low res texture feature only affects textures for the environment and fullscreen images; it doesn't affect texture resolution for things like characters and enemies.

mercury501 commented:
> I fixed the issue with low res textures getting re-enabled. I also added support for dynamically changing display modes on the fly and dynamically changing the health indicator.
> 
> @mercury501, I still see two "TODO" notes. [Here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/pull/935/files#diff-9c12d0c188c16765dde0cd9c23c19b1957a148784fb5a24adffce9d5f4b67235R1070) and [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/pull/935/files#diff-9c12d0c188c16765dde0cd9c23c19b1957a148784fb5a24adffce9d5f4b67235R1579). See below. What is the status of these two items?
> 
> ```c++
> void PatchHealthIndicatorOption() // TODO find faulty address in 1.1 and DC
> ```
> 
> ```c++
> WriteJMPtoMemory(DisplayModeValueHighlightAddr, DisplayModeValueHighlight, 0x08); //TODO highlight wrong color, tied to inputs
> ```

IIRC those should be taken care of, I just forgot to remove the todos, I'll check on it in a bit

mercury501 commented:
@elishacloud @Polymega Should be good now, let me know if you find any issue!

Polymega commented:
Did some quick tests for the health indicator feature on all binaries and looks to be working just well now! Thank you, mercury!

elishacloud commented:
@AeroWidescreen, would you be able to look up something for me?  I am trying to update the Display Mode feature in the Advanced Menu.  Currently, I am sending a device lost message to the game that tells the game to reset the display so then I can put the display into the different mode.  However, this has some side effects.

Instead, I would like the game to think that I am asking for a resolution change instead.  What I would like to do is make it so that when the Display Mode changes in the Advanced Menu the game thinks the resolution is changing and run the code that it would run on resolution change.

Let me know if you are able to help me find the code that tells the game that the resolution has change in the Advanced Menu.

elishacloud commented:
Thanks @mercury501!  Great work!
